scipy/optimize/_numdiff.py

```diff
@@ -0,0 +1,297 @@
+    """Adjust a step for the forward finite difference derivative estimation,
+    such that the perturbed point is inside the bounds.
+
+    This function tries to select -h[i] or +h[i] as a step for i-th variable.
+    If both such steps move the point out of the boundaries, then the largest
+    allowed step is selected with positive or negative sign.
+
+    Parameters
+    ----------
+    x0 : ndarray
+        Initial point. Assumed to lie inside the bounds, no checks of that
+        are done.
+    h : ndarray
+        Desired positive step sizes. Might be decreased in extreme cases.
+    bounds : tuple of array_like
+        Lower and upper bounds on the variables. None means no bounds.
+
+    Returns
+    -------
+    h_adjusted : ndarray
+        Adjusted steps to use in derivative estimation.
+    """
+    lb, ub = bounds
+    if ub is None:
+        return h
+    if lb is None:
+        return -h
+
+    lb = np.asarray(lb)
+    ub = np.asarray(ub)
+
+    lower_dist = x0 - lb
+    upper_dist = ub - x0
+    forward = upper_dist > lower_dist
+
+    h_adjusted = np.empty_like(h)
+    h_adjusted[forward] = np.minimum(upper_dist[forward], h[forward])
+    h_adjusted[~forward] = -np.minimum(lower_dist[~forward], h[~forward])
+
+    return h_adjusted
+
+
+def _adjust_step_2nd_order(x0, h, bounds):
+    """Select a finite difference scheme and adjust the step size for the
+    first derivative estimation with the second order accuracy, such that
+    perturbed points are inside the bounds.
+
+    This function chooses between central finite difference and 3-point
+    forward (backward) finite difference. Both methods provide second order
+    accuracy.Central difference scheme is chosen with higher priority.
+
+    Parameters
+    ----------
+    x0 : ndarray
+        Initial point. Assumed to lie inside the bounds, no checks of that
+        are done.
+    h : ndarray
+        Desired positive step sizes. Might be decreased in extreme cases.
+    bounds : tuple of array_like
+        Lower and upper bounds on the variables. None means no bounds.
+
+    Returns
+    -------
+    h_adjusted : ndarray
+        Adjusted steps to use in derivative estimation.
+    scheme : ndarray
+        Array containing 0 and 1. If 0 then use central finite difference,
+        if 1 then use 3-point forward finite difference.
+    """
+    lb, ub = bounds
+    if lb is None and ub is None:
+        return h, np.zeros_like(h, dtype=int)
+
+    if lb is None:
+        lb = np.empty_like(h)
+        lb.fill(-np.inf)
+
+    if ub is None:
+        ub = np.empty_like(h)
+        ub.fill(np.inf)
+
+    scheme = np.zeros_like(h, dtype=int)
+    h_adjusted = h.copy()
+
+    lower_dist = x0 - lb
+    upper_dist = ub - x0
+
+    central = (lower_dist >= h) & (upper_dist >= h)
+
+    forward = ~central & (upper_dist >= 2 * h)
+    scheme[forward] = 1
+    h_adjusted[forward] = h[forward]
+
+    backward = ~central & (lower_dist >= 2 * h)
```

Comment:
I might be wrong on this one, but it seems to me that these `_adjust_step{1st, 2nd}_order` are more like selecting even/odd ordered schemes. Then it's better for forward compat (and clarity) to add an extra argument to specify the max length of a stencil (either an integer `num` as in `num*h` or `max_step`).